### PR TITLE
Added support for cloudflare API tokens.

### PR DESCRIPTION
**Pull request by TheraNinjaCat**
I have added support for using cloudflare API tokens as opposed to a global key. It will decide which to use depending on if the login parameter is provided, if it is not, then it will attempt to use the password as an API token, else it uses it like a global key.

### CODE REVIEW

File: sample-etc_ddclient.conf [hunk 1]
```diff
@@ -210,8 +210,8 @@ ssl=yes					# use ssl-support.  Works with
 #protocol=cloudflare,        \
 #zone=domain.tld,            \
 #ttl=1,                      \
-#login=your-login-email,     \
-#password=APIKey             \
+#login=your-login-email,     \ # if using global key
```
Comment (SuperSandro2000) [inline sample-etc_ddclient.conf]:
```suggestion
#login=your-login-email,     \ # Only needed when you want to use a global api key.
```
I am not a native speaker so if someone has an even better idea feel free to improve me. :)

File: sample-etc_ddclient.conf [hunk 2]
```diff
@@ -210,8 +210,8 @@ ssl=yes					# use ssl-support.  Works with
...
+#password=APIKey             \ # if using an api token it must have permissions "All zones - Zone:Read, DNS:Edit"
```
Comment (SuperSandro2000) [inline sample-etc_ddclient.conf]:
```suggestion
#password=APIKey             \ # Only needed when you want to use an api token. It must have those permissions "Zone - DNS - Edit" and "Zone - Zone - Read" and Zone resources needs to include "All zones".
```

File: ddclient [hunk 3]
```diff
@@ -4522,25 +4523,31 @@ o 'cloudflare'
 The 'cloudflare' protocol is used by DNS service offered by www.cloudflare.com.
 
 Configuration variables applicable to the 'cloudflare' protocol are:
-  protocol=cloudflare          ## 
+  protocol=cloudflare          ##
   server=fqdn.of.service       ## defaults to api.cloudflare.com/client/v4
-  login=service-login          ## login name and password  registered with the service
-  password=service-password    ##
+  login=service-login          ## login name if using global key
```
Comment (SuperSandro2000) [inline ddclient]:
```suggestion
  login=service-login          ## login email when using a global api key
```

File: ddclient [hunk 4]
```diff
@@ -4522,25 +4523,31 @@ o 'cloudflare'
...
+  password=service-password    ## global key, or api token with at least the permissions "All zones - Zone:Read, DNS:Edit"
```
Comment (SuperSandro2000) [inline ddclient]:
```suggestion
  password=service-password    ## Global api key or api token with those permissions "Zone - DNS - Edit" and "Zone - Zone - Read" and Zone resources needs to include "All zones".
```

File: ddclient [hunk 5]
```diff
@@ -4522,25 +4523,31 @@ o 'cloudflare'
...
   fully.qualified.host         ## the host registered with the service.
 
 Example ${program}.conf file entries:
-  ## single host update
+  ## single host update using global key
```
Comment (SuperSandro2000) [inline ddclient]:
```suggestion
  ## single host update using global api key
```

File: ddclient [hunk 6]
```diff
@@ -4522,25 +4523,31 @@ o 'cloudflare'
...
   login=my-cloudflare.com-login,                               \\
-  password=my-cloudflare.com-secure-token                      \\
-  myhost.com 
+  password=my-cloudflare.com-global-key                        \\
+  myhost.com
+
+  ## single host update using api token
+  protocol=cloudflare,                                         \\
+  zone=dns.zone,                                               \\
+  password=my-cloudflare.com-api-token                         \\
```
Comment (SuperSandro2000) [inline ddclient]:
```suggestion
  password=cloudflare-api-token                                \\
```

File: ddclient [hunk 7]
```diff
@@ -4522,25 +4523,31 @@ o 'cloudflare'
...
+  myhost.com
 
   ## multiple host update to the custom DNS service
   protocol=cloudflare,                                         \\
   zone=dns.zone,                                               \\
   login=my-cloudflare.com-login,                               \\
-  password=my-cloudflare.com-secure-token                      \\
+  password=my-cloudflare.com-global-key                        \\
```
Comment (SuperSandro2000) [inline ddclient]:
```suggestion
  password=my-cloudflare-global-api-key                        \\
```

**Comment by SuperSandro2000** [comment]
some thoughts:
It would have been nice if you would have split formatting and code changes in two commits as this would make the review process quite a bit easier.
Also I nit picked a bit on the comments 😉

**Comment by TheraNinjaCat** [comment]
Thank you for your suggestions, I’ll implement those when I get a chance tomorrow morning. My excuse for the poor grammar is that I’m lazy and it’s a bad habit 😅
Sorry about the single commit with code and formatting changes, in all honesty I didn’t even notice the formatting changes until I had already done the commit and pull request as my editor has done that automatically without me noticing...

**Comment by SuperSandro2000** [comment]
JFYI:
I always use ``git commit -p`` to select the hunks I want to commit. Maybe this helps you in the future.

**Comment by alivahabi2112** [comment]
در تاریخ شنبه ۲۲ فوریهٔ ۲۰۲۰،‏ ۵:۵۰ Sandro <notifications@github.com> نوشت:

**Comment by R-Adrian** [comment]
FYI this API Token seems to be now mandatory with Cloudflare.
The old Global API Key login method now results in `HTTP/1.1 405 Method Not Allowed` server responses from the Cloudflare API and only API Tokens updates seem to work,

https://community.cloudflare.com/t/unable-to-update-ddns-using-api-for-some-tlds/167228/23

**Comment by SuperSandro2000** [comment]
We probably need to update the docs with that.

**Comment by Tiemichael** [comment]
I tried the new approach a couple of hours, ago and am not able to get if working.
The debug shows:
`DEBUG:    url    = https://api.cloudflare.com/client/v4/zones/89de6f41d90716f308a2079776f60fe5/dns_records?type=A&name=\
DEBUG:    server = api.cloudflare.com
CONNECT:  api.cloudflare.com
CONNECTED:  using SSL
SENDING:  GET /client/v4/zones/89de6f41d90716f308a2079776f60fe5/dns_records?type=A&name=\ HTTP/1.0
SENDING:   Host: api.cloudflare.com
SENDING:   User-Agent: ddclient/3.9.1
SENDING:   Connection: close
SENDING:   Content-Type: application/json
SENDING:   Authorization: Bearer nddddddYgVcj5tFexxxxxxxxM0cPeIQJ4eQYYR91
` and 
`RECEIVE:  Server: cloudflare
RECEIVE:  cf-request-id: 0252b697cb0000dcfa4826a200000001
RECEIVE:  
RECEIVE:  {"result":[],"result_info":{"page":1,"per_page":20,"total_pages":0,"count":0,"total_count":0},"success":true,"errors":[],"messages":[]}
FAILED:   updating \: No DNS record ID found.
`
If I compare the GET-line with the API-Documentation, it seems that the DNS-Record Name (example.com) is missing.

The A-record is defined as xxxxxxxx.com.

Thanks for the tool, and your help!

**Comment by SuperSandro2000** [comment]
@Tiemichael can you please open a new issue with that? otherwise we might lose it.

**Comment by R-Adrian** [comment]
1) make sure you also set the ACL for the Clouflare API Token on their site:
- All zones in account- Zone:Read,
- DNS:Edit

2) my ddclient.conf looks like this:
```
use=web, \
web=whatismyip.akamai.com \
ssl=yes, \
protocol=cloudflare, \
ttl=1, \
login=token, \
zone=example.com, \
password=my API Token here \
example.com, www.example.com
```

and it works for me.
```
RECEIVE:    "success": true,
RECEIVE:    "errors": [],
RECEIVE:    "messages": []

```

**Comment by kapotik** [comment]
Hello thanks for this contribution, can someone tell me how I update this version? I have installed the version: "ddclient 3.8.3-1.1ubuntu1"
Thank you

**Comment by SuperSandro2000** [comment]
Just grab the script from master and run it https://github.com/ddclient/ddclient/blob/master/ddclient

**Comment by kapotik** [comment]
sorry, newbie to linux, could be more specific, i already have installed a previous version of ddclient, what do i do with it, how do i install the new version?

**Comment by SuperSandro2000** [comment]
I think there are two ways to achieve this:
- overwrite the currently installed ddclient which is not really recommended but the fastest method. When your ddclient package gets an update it would be overwritten and you would need to repeat the steps if it is not new enough.
```
wget https://github.com/ddclient/ddclient/blob/master/ddclient
chmod +x ddclient
mv ddclient $(which ddclient)
```
- or you could download the script to a new location like ``/usr/local/bin/`` and update the path in systemd/initrc/sysv/cron/etc. This is probably a bit more advanced. 
- third and probably easiest without breaking recommendation: you download the script to a known location, add +x with chmod and add it to crontab.

I hope this helps you in finding the right information with the help of the internet.

**Comment by kapotik** [comment]
Thank you very much, I think with this I can achieve it!!!

**Comment by kapotik** [comment]
Hello again with this topic, I already have updated ddclient to vercion 3.9.1 but I still cannot update in cloudflare, I am sure that something is wrong, I leave my ddclient.conf:
`# Configuration file for ddclient generated by debconf
#
# /etc/ddclient.conf

use=web, \
web=whatismyip.akamai.com \
ssl=yes, \
protocol=cloudflare, \
ttl=1, \
login=token, \
zone=kapotik.tk, \
password=7GHJ5NxxxxelvBYiD0rNZ-s39w5-NZ7X38cxrxgq \
kapotik.tk, www.kapotik.tk`

and this is what throws me:
`CONNECT:  whatismyip.akamai.com
CONNECTED:  using HTTP
SENDING:  GET / HTTP/1.0
SENDING:   Host: whatismyip.akamai.com
SENDING:   User-Agent: ddclient/3.9.1
SENDING:   Connection: close
SENDING:
SENDING:
RECEIVE:  HTTP/1.0 200 OK
RECEIVE:  Content-Type: text/html
RECEIVE:  Content-Length: 13
RECEIVE:  Expires: Thu, 30 Apr 2020 01:15:56 GMT
RECEIVE:  Cache-Control: max-age=0, no-cache, no-store
RECEIVE:  Pragma: no-cache
RECEIVE:  Date: Thu, 30 Apr 2020 01:15:56 GMT
RECEIVE:  Connection: close
RECEIVE:
RECEIVE:  191.83.157.45
INFO:     forcing updating kapotik.tk because no cached entry exists.
INFO:     forcing updating www.kapotik.tk because no cached entry exists.
INFO:     setting IP address to 191.83.157.45 for kapotik.tk
UPDATE:   updating kapotik.tk
CONNECT:  api.cloudflare.com
CONNECTED:  using SSL
SENDING:  GET /client/v4/zones?name=kapotik.tk HTTP/1.0
SENDING:   Host: api.cloudflare.com
SENDING:   User-Agent: ddclient/3.9.1
SENDING:   Connection: close
SENDING:   Content-Type: application/json
SENDING:   Authorization: Bearer 7GHJ5Nxo7OelvBYiD0rNZ-s39w5-NZ7X38cxrxgq
SENDING:
RECEIVE:  HTTP/1.1 200 OK
RECEIVE:  Date: Thu, 30 Apr 2020 01:15:59 GMT
RECEIVE:  Content-Type: application/json
RECEIVE:  Connection: close
RECEIVE:  Set-Cookie: __cfduid=dd1b2a61105587b35b45813d930cb456f1588209358; expires=Sat, 30-May-20 01:15:58 GMT; path=/; domain=.api.cloudflare.com; HttpOnly; SameSite=Lax
RECEIVE:  CF-Ray: 58bd6a2d99f6677d-EZE
RECEIVE:  Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
RECEIVE:  Expires: Sun, 25 Jan 1981 05:00:00 GMT
RECEIVE:  Strict-Transport-Security: max-age=31536000
RECEIVE:  CF-Cache-Status: DYNAMIC
RECEIVE:  Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
RECEIVE:  Pragma: no-cache
RECEIVE:  X-Content-Type-Options: nosniff
RECEIVE:  X-Frame-Options: SAMEORIGIN
RECEIVE:  Server: cloudflare
RECEIVE:  cf-request-id: 026a40b0830000677dad90c200000001
RECEIVE:
RECEIVE:  {"result":[{"id":"95a2317d8787a471b3f0a0203a49f7fb","name":"kapotik.tk","status":"active","paused":false,"type":"full","development_mode":0,"name_servers":["hans.ns.cloudflare.com","karsyn.ns.cloudflare.com"],"original_name_servers":["ns04.freenom.com","ns01.freenom.com","ns02.freenom.com","ns03.freenom.com"],"original_registrar":null,"original_dnshost":null,"modified_on":"2020-04-25T00:23:21.336751Z","created_on":"2020-04-12T16:26:45.359417Z","activated_on":"2020-04-12T16:35:24.094296Z","meta":{"step":2,"wildcard_proxiable":false,"custom_certificate_quota":0,"page_rule_quota":3,"phishing_detected":false,"multiple_railguns_allowed":false},"owner":{"id":"998ec649cf16982ac89b7c8b65634925","type":"user","email":"energst@gmail.com"},"account":{"id":"418c3bd59dd6422b42a43aadfd365d08","name":"energst@gmail.com"},"permissions":["#access:edit","#access:read","#analytics:read","#app:edit","#auditlogs:read","#billing:edit","#billing:read","#cache_purge:edit","#dns_records:edit","#dns_records:read","#lb:edit","#lb:read","#legal:edit","#legal:read","#logs:edit","#logs:read","#member:edit","#member:read","#organization:edit","#organization:read","#ssl:edit","#ssl:read","#stream:edit","#stream:read","#subscription:edit","#subscription:read","#teams:edit","#teams:read","#teams:report","#waf:edit","#waf:read","#webhooks:edit","#webhooks:read","#worker:edit","#worker:read","#zone:edit","#zone:read","#zone_settings:edit","#zone_settings:read"],"plan":{"id":"0feeeeeeeeeeeeeeeeeeeeeeeeeeeeee","name":"Free Website","price":0,"currency":"USD","frequency":"","is_subscribed":false,"can_subscribe":false,"legacy_id":"free","legacy_discount":false,"externally_managed":false},"plan_pending":{"id":"0feeeeeeeeeeeeeeeeeeeeeeeeeeeeee","name":"Free Website","price":0,"currency":"USD","frequency":"","is_subscribed":false,"can_subscribe":false,"legacy_id":"free","legacy_discount":false,"externally_managed":false}}],"result_info":{"page":1,"per_page":20,"total_pages":1,"count":1,"total_count":1},"success":true,"errors":[],"messages":[]}
INFO:     zone ID is 95a2317d8787a471b3f0a0203a49f7fb
CONNECT:  api.cloudflare.com
CONNECTED:  using SSL
SENDING:  GET /client/v4/zones/95a2317d8787a471b3f0a0203a49f7fb/dns_records?type=A&name=kapotik.tk HTTP/1.0
SENDING:   Host: api.cloudflare.com
SENDING:   User-Agent: ddclient/3.9.1
SENDING:   Connection: close
SENDING:   Content-Type: application/json
SENDING:   Authorization: Bearer 7GHJ5Nxo7OelvBYiD0rNZ-s39w5-NZ7X38cxrxgq
SENDING:
RECEIVE:  HTTP/1.1 200 OK
RECEIVE:  Date: Thu, 30 Apr 2020 01:15:59 GMT
RECEIVE:  Content-Type: application/json; charset=UTF-8
RECEIVE:  Connection: close
RECEIVE:  Set-Cookie: __cfduid=d35200716e04047c7e38e121dd1bcba8f1588209359; expires=Sat, 30-May-20 01:15:59 GMT; path=/; domain=.api.cloudflare.com; HttpOnly; SameSite=Lax
RECEIVE:  CF-Ray: 58bd6a31e88267bf-EZE
RECEIVE:  Vary: Accept-Encoding
RECEIVE:  CF-Cache-Status: DYNAMIC
RECEIVE:  Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
RECEIVE:  x-envoy-upstream-service-time: 3
RECEIVE:  Server: cloudflare
RECEIVE:  cf-request-id: 026a40b331000067bfa81d7200000001
RECEIVE:
RECEIVE:  {
RECEIVE:    "result": [
RECEIVE:      {
RECEIVE:        "id": "b05f3229b56ad8f2ca41e848f4db3e62",
RECEIVE:        "zone_id": "95a2317d8787a471b3f0a0203a49f7fb",
RECEIVE:        "zone_name": "kapotik.tk",
RECEIVE:        "name": "kapotik.tk",
RECEIVE:        "type": "A",
RECEIVE:        "content": "191.83.159.221",
RECEIVE:        "proxiable": true,
RECEIVE:        "proxied": false,
RECEIVE:        "ttl": 1,
RECEIVE:        "locked": false,
RECEIVE:        "meta": {
RECEIVE:          "auto_added": false,
RECEIVE:          "managed_by_apps": false,
RECEIVE:          "managed_by_argo_tunnel": false,
RECEIVE:          "source": "primary"
RECEIVE:        },
RECEIVE:        "created_on": "2020-04-25T00:23:21.336751Z",
RECEIVE:        "modified_on": "2020-04-25T00:23:21.336751Z"
RECEIVE:      }
RECEIVE:    ],
RECEIVE:    "success": true,
RECEIVE:    "errors": [],
RECEIVE:    "messages": [],
RECEIVE:    "result_info": {
RECEIVE:      "page": 1,
RECEIVE:      "per_page": 20,
RECEIVE:      "count": 1,
RECEIVE:      "total_count": 1,
RECEIVE:      "total_pages": 1
RECEIVE:    }
RECEIVE:  }
INFO:     DNS record ID is b05f3229b56ad8f2ca41e848f4db3e62
CONNECT:  api.cloudflare.com
CONNECTED:  using SSL
SENDING:  PATCH /client/v4/zones/95a2317d8787a471b3f0a0203a49f7fb/dns_records/b05f3229b56ad8f2ca41e848f4db3e62 HTTP/1.0
SENDING:   Host: api.cloudflare.com
SENDING:   User-Agent: ddclient/3.9.1
SENDING:   Content-Length: 27
SENDING:   Connection: close
SENDING:   Content-Type: application/json
SENDING:   Authorization: Bearer 7GHJ5Nxo7OelvBYiD0rNZ-s39w5-NZ7X38cxrxgq
SENDING:
SENDING:   {"content":"191.83.157.45"}
RECEIVE:  HTTP/1.1 401 Unauthorized
RECEIVE:  Date: Thu, 30 Apr 2020 01:16:00 GMT
RECEIVE:  Content-Type: application/json; charset=UTF-8
RECEIVE:  Connection: close
RECEIVE:  Set-Cookie: __cfduid=de81626676d24f1ef9de659d8f624bf671588209360; expires=Sat, 30-May-20 01:16:00 GMT; path=/; domain=.api.cloudflare.com; HttpOnly; SameSite=Lax
RECEIVE:  CF-Ray: 58bd6a36fd2dd7f9-EZE
RECEIVE:  Vary: Accept-Encoding
RECEIVE:  CF-Cache-Status: DYNAMIC
RECEIVE:  Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
RECEIVE:  x-envoy-upstream-service-time: 1
RECEIVE:  Server: cloudflare
RECEIVE:  cf-request-id: 026a40b6590000d7f911b4d200000001
RECEIVE:
RECEIVE:  {
RECEIVE:    "error": "You cannot use this API for domains with a .cf, .ga, .gq, .ml, or .tk TLD (top-level domain). To configure the DNS settings for this domain, use the Cloudflare Dashboard."
RECEIVE:  }
FAILED:   updating kapotik.tk: authorization failed (HTTP/1.1 401 Unauthorized
FAILED:    Date: Thu, 30 Apr 2020 01:16:00 GMT
FAILED:    Content-Type: application/json; charset=UTF-8
FAILED:    Connection: close
FAILED:    Set-Cookie: __cfduid=de81626676d24f1ef9de659d8f624bf671588209360; expires=Sat, 30-May-20 01:16:00 GMT; path=/; domain=.api.cloudflare.com; HttpOnly; SameSite=Lax
FAILED:    CF-Ray: 58bd6a36fd2dd7f9-EZE
FAILED:    Vary: Accept-Encoding
FAILED:    CF-Cache-Status: DYNAMIC
FAILED:    Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
FAILED:    x-envoy-upstream-service-time: 1
FAILED:    Server: cloudflare
FAILED:    cf-request-id: 026a40b6590000d7f911b4d200000001
FAILED:
FAILED:    {
FAILED:      "error": "You cannot use this API for domains with a .cf, .ga, .gq, .ml, or .tk TLD (top-level domain). To configure the DNS settings for this domain, use the Cloudflare Dashboard."
FAILED:    }
FAILED:    )
`

**Comment by kapotik** [comment]
so configure the token in cloudflare:
cloud

Edit zone DNS API token summary
This API token will affect the below accounts and zones, along with their respective permissions


All zones - Zone:Read, DNS:Edit

**Comment by TheraNinjaCat** [comment]
In your configuration you don't actually need to specify login at all, as it'll default to `token`, but that doesn't really matter.

The answer to your question is in the response from the api: `RECEIVE: HTTP/1.1 401 Unauthorized`
`"error": "You cannot use this API for domains with a .cf, .ga, .gq, .ml, or .tk TLD (top-level domain). To configure the DNS settings for this domain, use the Cloudflare Dashboard."`

This suggests to me that Cloudflare have disallowed modifying these TLDs from their API, likely as a way to limit usage. Therefore you won't be able to use ddclient for the domain `kapotik.tk` through Cloudflare.

Also, you appear to have partially hidden your token with x's here `password=7GHJ5NxxxxelvBYiD0rNZ-s39w5-NZ7X38cxrxgq` however in the logs you posted it contains the bearer token `SENDING: Authorization: Bearer 7GHJ5Nxo7OelvBYiD0rNZ-s39w5-NZ7X38cxrxgq`. I would recommend you delete this token and generate a new one, as this one is now compromised. Also, replacing 3 x's in there is barely hiding it at all.

**Comment by kapotik** [comment]
thanks for your reply, i understand what you are saying but is not that this new version of ddclient 3.9.1 has a patch to be able to update through the cloudflare api domains of type .tk?

**Comment by NHAS** [comment]
@kapotik Your account tokens for updating your dns records are now shared with anyone who reads this. 
You need to change your token ASAP

**Comment by kapotik** [comment]
ok changed

**Comment by NHAS** [comment]
In short, anyone can now take over kapotik.tk if you dont change your token @kapotik

**Comment by TheraNinjaCat** [comment]
@kapotik the patch I did to ddclient did nothing for `.tk` domains, it simply enabled using their limited API tokens instead of the global API key. The limitation around the TLD is on Cloudflare's end and no patch to ddclient can make their API accept them.

**Comment by kapotik** [comment]
Ok, thanks for your answers, I was confused regarding the ddclient update

**Comment by SuperSandro2000** [comment]
I replaced some sensitive information but they are probably still leaked somewhere and you should reset the password and tokens asap.

**Comment by kapotik** [comment]
Thank you. It is no longer an option for my cloudflare so I canceled the account, there is no problem for the token exposure

**Comment by FingerlessGlov3s** [comment]
Not sure if this is the right place, but when will a new version of ddclient get released with this PR?

**Comment by SuperSandro2000** [comment]
When I get to release it but that will take a few weeks.

**Comment by SuperSandro2000** [comment]
https://github.com/ddclient/ddclient/releases/tag/v3.10.0_2

**Pull request merged**